lib/events/stream.go

```diff
@@ -376,6 +394,11 @@ func (s *ProtoStream) EmitAuditEvent(ctx context.Context, event apievents.AuditE
 		}
 		return nil
 	case <-s.cancelCtx.Done():
+		cancelErr := s.getCancelError()
+		if cancelErr != nil {
+			return trace.Wrap(cancelErr)
+		}
+
 		return trace.ConnectionProblem(s.cancelCtx.Err(), "emitter has been closed")
```

Comment:
Capturing the `cancelError` would actually be really nice for the other errors too - I can't count how many times I've seen "emitter has been closed" and had to dig up why it was closed.

How about updating it like this? This would clean up the logging and context canceling, and improve the error message for future calls to `EmitAuditEvent`.
```
go func() {
    err := writer.receiveAndUpload()
    if err != nil {
        log.WithError(err).Debug("slice writer ended with error")
        writer.proto.SetCancelError(err)
    }
    a.cancel()
}
```

Reply:
Done. Thanks for the suggestion.